### PR TITLE
Internal FeedIterator: Refactors code path for FeedIterators through clientcontext.OperationHelper

### PR DESCRIPTION
**Pull request by asketagarwal**
# Pull Request Template

## Description

This PR utilizes the ClientContext.OperationHelperAsync for recording the trace for all public api's using FeedIterators's
Also adds a perf test for Diagnostics.ToString()

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 1]
```diff
@@ -821,7 +821,8 @@ public virtual Container GetContainer(string databaseId, string containerId)
                this.GetDatabaseQueryIteratorHelper<T>(
                    queryDefinition,
                    continuationToken,
-                   requestOptions));
+                   requestOptions),
+               this.ClientContext);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:825]: nit: fix indenting

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerInlineCore.cs [hunk 2]
```diff
@@ -347,7 +350,8 @@ Task<ResponseMessage> func(ITrace trace)
             return new FeedIteratorInlineCore<T>(base.GetItemQueryIterator<T>(
                 queryText,
                 continuationToken,
-                requestOptions));
+                requestOptions),
+                this.ClientContext);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerInlineCore.cs:354]: What about ChangeFeed pull model?

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs [hunk 3]
```diff
@@ -110,92 +113,103 @@ public virtual bool TryAdd(ItemBatchOperation operation)
...
-            {
-                this.interlockIncrementCheck.EnterLockCheck();
+            await this.clientContext.OperationHelperAsync("Batch Dispatch Async",
+                        requestOptions: null,
+                        task: (trace) => this.DispatchAsync(trace, partitionMetric, cancellationToken),
+                        traceComponent: TraceComponent.Batch,
+                        traceLevel: Tracing.TraceLevel.Info);
+        }
+
+        public virtual async Task<object> DispatchAsync(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncBatcher.cs:123]: private

File: Microsoft.Azure.Cosmos/src/Resource/CosmosClientContext.cs [hunk 4]
```diff
@@ -60,7 +60,9 @@ internal abstract class CosmosClientContext : IDisposable
         internal abstract Task<TResult> OperationHelperAsync<TResult>(
             string operationName,
             RequestOptions requestOptions,
-            Func<ITrace, Task<TResult>> task);
+            Func<ITrace, Task<TResult>> task,
+            TraceComponent traceComponent = TraceComponent.Transport,
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosClientContext.cs:64]: The default should be unkown.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/BatchAsyncContainerExecutorTests.cs [hunk 5]
```diff
@@ -28,7 +28,7 @@ public async Task RetryOnSplit()
         {
             ItemBatchOperation itemBatchOperation = CreateItem("test");
 
-            Mock<CosmosClientContext> mockedContext = new Mock<CosmosClientContext>();
+            Mock<CosmosClientContext> mockedContext = this.GetMockClientContext();
```
Comment (bartelink) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/BatchAsyncContainerExecutorTests.cs:31]: suggestion: Get -> Create (or just remove the Get)

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Benchmarks/ClientConfigurationTraceDatumBenchmark.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+        [Benchmark]
+        [BenchmarkCategory("GateBenchmark")]
+        public async Task ReadItemTDiagnostics()
+        {
+            ItemResponse<ToDoActivity> response = await this.benchmarkHelper.TestContainer.ReadItemAsync<ToDoActivity>(
+                            MockedItemBenchmarkHelper.ExistingItemId,
+                            MockedItemBenchmarkHelper.ExistingPartitionId);
+
+            response.Diagnostics.ToString();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Benchmarks/ClientConfigurationTraceDatumBenchmark.cs:28]: This should only benchmark the .ToString(); The operation should be done as part of the initialization. We only want to see the overhead added to getting the string. Having the item operation will cause it make it difficult to tell if the change is from the diagnostics or the item operaiton.

**Review by j82w — APPROVED** [review]

**Pull request merged**